core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -788,7 +788,9 @@ public PlanOptimizers(
                         ruleStats,
                         statsCalculator,
                         costCalculator,
-                        ImmutableSet.of(new ReorderJoins(plannerContext, costComparator))));
+                        ImmutableSet.of(new ReorderJoins(plannerContext, costComparator))),
+                // ReorderJoins may produce filters above joins that could (and should be) pushed back down
+                new StatsRecordingPlanOptimizer(optimizerStats, new PredicatePushDown(plannerContext, true, false)));
```

Comment:
This is probably of, but it masks the underlying problem. Namely, that the connector is not handling `filter(<f>, innerjoin(l, r, true))` and `innerjoin(l, r, <f>)` consistently. 

If there are connector tests relying on that, the connector should be fixed, as there's is not guarantee that the applyXXX calls will be made at any specific point during the planning process.

Reply:
> This is probably of,

looks truncated

> that the connector is not handling filter(<f>, innerjoin(l, r, true)) and innerjoin(l, r, <f>) consistently.

To be precise, `innerjoin(l, r, true)` is not pushed down into connectors at all today (https://github.com/trinodb/trino/blob/90ef6e5f3d40d89de34e8e2b20483624d53c5250/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java#L100)
so a better example would be `filter(<f>, innerjoin(l, r, <j>))` and `innerjoin(l, r, <f> AND <j>)` differently.

it is not a bug in itself. The connector must handle them differently if `<f>` is a predicate the connector does not understand / does not support.

Reply:
> looks truncated

Oops. It should be "probably ok"

Reply:
> Oops. It should be "probably ok"

so nice, thanks!

> it masks the underlying problem. Namely, that the connector is not handling filter(<f>, innerjoin(l, r, true)) and innerjoin(l, r, <f>) consistently.

Can my answer above be used to conclude that the connector cannot do much differently than it does today? (under current SPI model, of course)

Reply:
Yes, you're right. We'd need to change the SPI so that applyJoin could return a residual filter.

Reply:
correct, but not only that. We would need to extend `JoinStatistics` (part of `applyJoin()` call) to support estimations for subset of join conditions.